common/src/db/query.rs

```diff
@@ -17,38 +17,72 @@ use utoipa::IntoParams;
 // Public interface
 /////////////////////////////////////////////////////////////////////////
 
-#[derive(
-    Clone,
-    Default,
-    Debug,
-    serde::Deserialize,
-    serde::Serialize,
-    utoipa::ToSchema,
-    utoipa::IntoParams,
-)]
-#[serde(rename_all = "camelCase")]
-pub struct SearchOptions {
-    /// The search filter
-    #[serde(default)]
-    pub q: String,
-    #[serde(default)]
-    /// Sort options
-    pub sort: String,
+/// Convenience function for creating a search query
+///
+/// ```
+/// use trustify_common::db::query::q;
+///
+/// let query = q("foo&bar>100").sort("bar:desc");
+///
+/// ```
+pub fn q(s: &str) -> Query {
+    Query::q(s)
 }
 
-#[derive(Debug, thiserror::Error)]
-pub enum Error {
-    #[error("query syntax error: {0}")]
-    SearchSyntax(String),
+impl Query {
+    /// Form expected: `{query}*{filter}*`
+    ///
+    /// where `{filter}` is of the form `{field}{op}{value}`
+    ///
+    /// Multiple queries and/or filters should be `&`-delimited
+    ///
+    /// The `{query}` text will result in an OR clause of LIKE clauses
+    /// for every [String] field in the associated
+    /// [Entity](sea_orm::EntityTrait). Optional filters of the form
+    /// `{field}{op}{value}` may further constrain the results. Each
+    /// `{field}` must name an actual
+    /// [Column](sea_orm::EntityTrait::Column) variant.
+    ///
+    /// Both `{query}` and `{value}` may contain `|`-delimited
+    /// alternate values that will result in an OR clause. Any `|` or
+    /// `&` within a query/value should be escaped with a backslash,
+    /// e.g. `\|` or `\&`.
+    ///
+    /// `{op}` should be one of `=`, `!=`, `~`, `!~, `>=`, `>`, `<=`,
+    /// or `<`.
+    pub fn q(s: &str) -> Self {
```

Comment:
so if I call .sort() first, and then q(), my sort gets trampled by ""?

Reply:
"Don't do that" is an acceptable answer, maybe.

Reply:
`.sort()` takes a `self` but `.q()` doesn't. There's no trampling because there's a field for each. I just figured `q` was the more likely used field. You could always just `Query {sort: "foo".into(), q: "".into() }`. It's ugly, but it works.

I'm open to better ideas if you have 'em.

Reply:
Alternatively, if you only needed sorting, `q("").sort("foo")`

Reply:
We could also introduce a top-level `sort`, in the same way `query::q()` calls `query::Query::q()`. So if you find yourself wanting to only sort a lot, we could create `query::sort()`